### ISSUE TITLE
vNext / MVC 6 Integration

### ISSUE THREAD

**Issue by endeffects** [JeremySkinner/FluentValidation#98]
Hi,

i saw you have created a branch for vNext / Mvc6. When do you plan a beta / final release?

Regards

**Comment by JeremySkinner** [comment]
I am away with work until September, so I'm not going to have much chance to work on it until then I'm afraid. I expect mid/end September will be the earliest I can do an initial release (unless someone else wants to help out)

**Comment by endeffects** [comment]
Thanks for your quick response. Than i have tu put this into my backlog. :)

**Comment by ddotlic** [comment]
@JeremySkinner I might have some time to do this at work (win-win for everyone) since my employer needs this (we're starting our work on next major version of the product and we'll be depending on Fluent Validation).

Is the work to be done relatively easy to explain? The hooks in MVC 6 should be much cleaner than before, I recall a brief exchange with you a few months back happened exactly at the moment Microsoft finally responded to one of your ASP.NET MVC github issues on the subject of Fluent's integration with MVC6, you were very hopeful following that.

What can I do to help?

**Comment by JeremySkinner** [comment]
Thanks, I appreciate the offer. At the moment there's not much more that can be done...I'm still waiting for the asp.net team to make the following changes:

- Support for localization satellite assemblies in vnext projects
- Changes to the validation API to make it easier to add 3rd party validation libraries

Once these are done then I can do the following:

- Start using the vnext project file to generate outputs for all supported frameworks (rather than having separate projects for .net, vnext, pcl etc). I've done most of the work for this already, I just need to re-enable it 
- Add the MVC integration project. This is where most of the work will be - I haven't started on it at all because the API is still in flux. I don't quite know what it'll look like yet until I see the final validation API in MVC, but it should be quite different to the existing integration projects. 

I'll post an update here when I know more about what'll be needed.

**Comment by ddotlic** [comment]
> Changes to the validation API to make it easier to add 3rd party validation libraries

I was under (did not check) impression they've done it based on your feedback? Still not done? Sigh. It's quite late to be adding these kind of modifications (they say they'll be feature complete in the next beta which is due in 2-3 weeks time).

>Start using the vnext project file

Based on what I've seen elsewhere this is going to be a pain.

>Add the MVC integration project. This is where most of the work will be - I haven't started on it at all because the API is still in flux.

Shame that the API still hasn't stabilized.

Oh well, let's revisit this subject a bit later :smile:

**Comment by ddotlic** [comment]
Just looking at our previous exchange - issue https://github.com/aspnet/Mvc/issues/1260 is closed, with two other issues open of which one has already been closed leaving only one actually open: https://github.com/aspnet/Mvc/issues/2303

I've seen you commented on this just last day and they seem to planning on tackling this thing this week.

Is the 2303 really the last thing that needs to be done?

I'm personally more weary of the trouble with NuGet 3 (targeting the `dotnet` "platform" in the package) than with this.

**Comment by JeremySkinner** [comment]
Yes, I think 2303 is the last mvc issue I need changed. I'm also waiting for satellite assembly support which should also be coming with beta 8.

**Comment by ddotlic** [comment]
OK, good to know. Let's hope beta8 comes as planned and without any further delays. Cheers!

**Comment by JeremySkinner** [comment]
So beta8 is now out - I've updated the core project to build against beta8, and I'm pleased to say that localization support is now in. 

I haven't started on the mvc6 project yet, so if you still have some time to help out then that would be great!

**Comment by ddotlic** [comment]
We're just now working on integrating beta8 bits too at my employer's. In a few days I'll be free to work on whatever comes next. I do need to see with my employer because in my "personal" time there's very little free (transitioning from full time to consulting but already working on consulting and a 3yo at home...). Maybe we could already see how to split the work best? While I have extensive experience with coding in general and .NET in particular, I do not know your code base and might need a bit of time to adapt - is there anything sufficiently isolated for me to work on?

**Comment by JeremySkinner** [comment]
So the thing I'd particularly like help with is stubbing out what the new FluentValidationModelValidatorProvider should look like - if that's something that you happen to get a chance to look into it'd be very much appreciated. I haven't had a chance to look at the API since https://github.com/aspnet/Mvc/issues/2303 and https://github.com/aspnet/Mvc/issues/2304 have been implemented. 

I don't think this should particularly require any in-depth knowledge of FV's codebase - the only thing to remember is the major difference between FV and DataAnnotations, where FV needs to run after all properties have been set, whilst the DataAnnotations provider validates each value before the property is set.

**Comment by ddotlic** [comment]
I did not know that about DataAnnotations but yet another reason not to use them ;)

I'll leave myself a reminder to look into `FluentValidationModelValidatorProvider`

**Comment by jln1989** [comment]
@JeremySkinner  @ddotlic : Is Satellite assembly support is available now as Beta8 is released? I did not find any information regarding the same.

**Comment by JeremySkinner** [comment]
Yes, it is.

**Comment by ddotlic** [comment]
@JeremySkinner My employer works in 2 week sprints, we end this sprint in 2 days, I'll try to put in some planned time for fluent validation for next sprint and then look into what you've asked me to.
I'll take a look at the overall structure of the project (from the inside) in the meantime on my own time.

**Comment by jln1989** [comment]
@JeremySkinner : Can you point to any info regarding the same as it would be really helpful to me?

**Comment by JeremySkinner** [comment]
It just works out of the box...if  you have resource files in your project then satellite assemblies are generated at build time.

**Comment by JeremySkinner** [comment]
@ddotlic, thanks, I appreciate it

**Comment by ddotlic** [comment]
I managed to secure _some_ time for Fluent this sprint, but it's not clear yet how much. Unfortunately, companies are much more eager to use OSS work than to contribute to it. I'll keep fighting though :wink:

**Comment by janhartmann** [comment]
What is the status on this? I might can help out as well. 

Been a fan of FV for years and recently starting using .NET MVC 6, so I am looking forward to this is out. Let me know whats the status is and I would sure help out where I can.

**Comment by ddotlic** [comment]
@janhartmann  As usual, life happened (at least for me). Prior to today, had no personal time and my employer (which just became my client as I've gone independent) had other things to work on.

Just came back from vacation and if we're lucky, my client will have me work on validation aspects of the solution in question which should mean some time "free" to work on this.

Like @JeremySkinner due to my consulting obligations, I won't have time outside the normal working hours (already booked 120%, :tired_face:) but it's definitely time we get the ball rolling as ASP.NET 5 is in RC1 phase and things should be pretty stable by now (though based on my experience so far it's really not, Microsoft is still changing small things, enough to irritate us :wink: )

**Comment by SvenVandenbrande** [comment]
Let's hope you both have some time in the coming weeks to accomplish this. :)

**Comment by JeremySkinner** [comment]
Sorry I've not had any time to work on this at all recently. 

Some things I'm thinking about though:
I'm incredibly uncomfortable with using the "FluentValidation-vnext" project that's currently in the repo. My initial plan was to use this one project to produce multiple outputs targeting different platforms, which would help keep the project clean (ie only 1 codebase, rather than having several different projects/PCLs like at the moment. 

Unfortunately I've not had much luck with this due to the constant changes being made to the project.json format, and also the incredibly vague and unhelpful build errors that make problems incredibly difficult to track down. 

So instead I'm planning on keeping the solution in a similar format to it is now, with separate projects producing explicit outputs:

FluentValidation: The old core project will essentially remain as-is targetting 4.5 
FluentValidation.Portable: This will be expanded to target ASP.NET Core 5 
FluentValidation.Portable40: This will stay as-is, and is essentially deprecated now that MS no longer support .NET 4.0 
FluentValidation.MVC5: Stay as-is
FluentValidation.MVC6: New PCL to target ASP.NET Core 5
FluentValidation.Win10: New PCL to target universal Win10 

This shouldn't require too much work, as code changes for all this have already been done.

I haven't even begun on the MVC6 integration since beta8 was released. This is where the bulk of the work will be, and is still a long way off. I don't personally have any MVC6 projects in the pipeline (or in fact, any projects that use FV), which makes it very hard to find time to work on MVC6 FV integration. I can sometimes spend 1-2 hours per week on FV, at most.

**Comment by SvenVandenbrande** [comment]
I have some application that i am writing for the moment in ASP.NET 5 MVC 6, so no problem to test it..

**Comment by ddotlic** [comment]
> So instead I'm planning on keeping the solution in a similar format to it is now, with separate projects producing explicit outputs:

I am affraid that this is the only sensible course of action, however silly it might seem.

I agree with you 110% that Microsoft's message, thinking and recent attempts at specifying how to package code in the new world are nothing but confusing.

>  I don't personally have any MVC6 projects in the pipeline 

My ex-employer (now client) actually does, so we're in a very good position to test the integration on a real-life codebase.

I've recently integrated several aspects of a typical Web site into the code and MVC6 has been really, really great. There is the issue of extremely sparse documentation (for now) and API changes even in this phase, but overall I've enjoyed every step of the way.

Differently put, I have high hopes that FV will - this time - integrate in a much easier way than before.

**Comment by janhartmann** [comment]
I have some almost-live-projects on mvc6 which I can test with as well.

**Comment by JeremySkinner** [comment]
Quick update on this: 

- I've now moved the solution entirely over to VS2015 
- I've added a FluentValidation.Portable46 project. 

The Portable46 project is one of the new format PCL projects (using project.json) and targets .NET 4.6, Universal Win10 and ASP.NET Core 5.0 

Unfortunately, there's a CLR bug in RC1 that prevents referencing aspnet5 class libraries from the new format PCLs. This is supposedly due to be fixed in RC2: https://github.com/dotnet/coreclr/issues/1964 

The result of this is that I can't actually reference the MVC6 dlls at the moment, which makes writing the MVC integration rather difficult! 

So I guess the MVC integration is on hold until RC2 is released.

**Comment by janhartmann** [comment]
Thanks for the update - looking forward to RC2! :+1:

**Comment by sinanbozkus** [comment]
Thanks, we will wait :+1:

**Comment by jeremymeng** [comment]
Update in https://github.com/dotnet/coreclr/issues/1964 recommended updating the `dotnet` TFM to one of the netstandard TFM's.

**Comment by jeremymeng** [comment]
@JeremySkinner Would it be possible to release a NuGet package with the FluentValidation.Portable46 support so that downstream library owners can be unblocked?  I am on the .NET team and helping porting OSS libraries to .NET Core. Please let me know how I can help.

**Comment by JeremySkinner** [comment]
The portable46 build is actually already available in the latest release on
nuget, but is currently only in the uap10 folder....do you know what is the
correct TFM to use for aspnet core now? It's changed so much I haven't kept
up (and not really keen to look into it again until it stabilises).

Thanks

**Comment by jeremymeng** [comment]
In RC1 `dotnet` would make it compatible.  However it's changed.  I believe `netstandard1.3` would be the one in RC2 that supports net46, uap10, and dnxcore50.

**Comment by JeremySkinner** [comment]
Thanks, that's great. I'll get that changed for when rc2 comes out.

**Comment by RehanSaeed** [comment]
Please release the FluentValidation.MVC6 NuGet package, even a hacked together project will do for now, there is only [one class](https://github.com/JeremySkinner/FluentValidation/tree/88255e1fbd1970770b8d3ee5469cabeadb180fa1/src/FluentValidation.Mvc6) in the project so it's easy to do.

**Comment by JeremySkinner** [comment]
@RehanSaeed Aspnetcore is still too unstable and there is a lot of work left to do before I can release an MVC package. The APIs are constantly changing, even though it is labelled as "RC" there is still a huge amount of churn. You'll need to wait for now- porting FluentValidation to aspnetcore is a huge amount of work, and I can only do it in my spare time, which I don't have much of. Please be patient.

**Comment by RehanSaeed** [comment]
Agreed that it's unstable. I totally understand the time you put in, as I'm in the same boat. You can release the package as pre-release, it's what most library authors (myself included) have been doing. I'd be willing to put in a PR if you'd like.

**Comment by JeremySkinner** [comment]
I haven't even finished the design of what I want the MVC6 integration to look like yet...as I said, you'll need to wait for now.

**Comment by ddotlic** [comment]
I think we all agreed that due to the API churn and some last-moment changes we'll wait for RC2? Which in turn *should* be right around the corner and this time should be stable enough to build upon (contrary to RC1 which didn't).

**Comment by RehanSaeed** [comment]
I hadn't realized the filter was not ready, now that I look at the project.json, it seems to be using Beta 7 bits.

**Comment by JeremySkinner** [comment]
None of it is written yet....it's just not been stable enough to even begin. I attempted with Beta 7, and again with RC1, but ran into showstoppers both times.  Once RC2 is out I'll be able to make a start, but that will still take a while, so it won't be available immediately after RC2 is out.

**Comment by jeremymeng** [comment]
@JeremySkinner I did an experiment for FV Core using the `netstandard1.3` TFM in https://github.com/jeremymeng/FluentValidation/tree/netstandard. I basically copied the csproj from the Portable46 project, then added [the following at the end](https://github.com/jeremymeng/FluentValidation/blob/netstandard/src/FluentValidation.Portable.netstandard/FluentValidation.Portable.netstandard.csproj#L371-L373).
```xml
  <PropertyGroup>
    <NuGetTargetMoniker>.NETStandard,Version=v1.3</NuGetTargetMoniker>
  </PropertyGroup>
```
I also change the project.json to have `netstandard1.3`: https://github.com/jeremymeng/FluentValidation/blob/netstandard/src/FluentValidation.Portable.netstandard/project.json#L13.

I copied the nuget pack stuffs from the NET46 as well.  The generated .nupkg was tested working in a dotnet/cli `netcoreapp`: https://github.com/jeremymeng/cli-samples/commit/4cc963c45c49f4b83db300d99a37a30428179de0, using a recent dotnet/cli build.

I hope this info would be helpful.  I used the MSBuild-based approach (more info can be found at http://aka.ms/port2core, though `dotnet5.4` is still used there.)    Of course you could go with the `project.json`-based option if you like.

/cc @thecodejunkie @jchannon.

**Comment by JeremySkinner** [comment]
That's great- thanks for taking the time to research and document this.

**Comment by JeremySkinner** [comment]
@jeremymeng did you get a chance to try this with Xamarin at all? Last time I tried moving to the "dotnet" tfm, it broke Xamarin builds (they ended up picking this rather than the portable45 library). Do you know if this is still an issue with netstandrd?

**Comment by thecodejunkie** [comment]
@JeremySkinner @jeremymeng in what way does Xamarin Studio care about all of this? Do they have their own NuGet client? If they do then I expect it would be lagging behind a bit since `netstandard1.x` is just coming into use in the main repository (replacing `dotnet`).

**Comment by JeremySkinner** [comment]
I'm pretty sure this was an issue for Xamarin projects in VS rather than
Xamarin studio.....I don't have any of the Xamarin integration installed on
my machine so can't test it myself.

**Comment by jchannon** [comment]
Or do you mean dnx support in XS?
http://lastexitcode.com/blog/2016/01/01/DNXSupportInXamarinStudio/

**Comment by jeremymeng** [comment]
@JeremySkinner Do you mean that if a package supports both `portable-net45+wp80+win8+wpa81` and `dotnet`, the latter would be used when restoring for a Xamarin project?  If multiple target frameworks are compatible with the project I believe there's a precedence check to decide which one is "better".  Do you have details about what didn't work?

**Comment by JeremySkinner** [comment]
@jeremymeng Yes, that was what it looked like...it seemed to pick the dotnet build rather than the PCL, but then generated build errors. This is the link to the original issue that was reported: https://github.com/JeremySkinner/FluentValidation/issues/203

**Comment by JeremySkinner** [comment]
@jeremymeng Another question...does this need to be a separate project, or can we just pack the same binaries from the 4.6 PCL? I couldn't see any difference in the projects apart from the NuGetTargetMoniker

**Comment by thecodejunkie** [comment]
I'm not very comfortable around PCLs so the best I can offer right now is this mapping table https://github.com/dotnet/corefx/blob/master/Documentation/architecture/net-platform-standard.md#portable-profiles

That said, I am here to offer my help. As you might now I'm currently on paid work to get Nancy running on Core. We're progressing a long nicely but I know that my client also depend on FluentValidation (preferably ready for rc2 when the stable bits comes out in 3-6 weeks) so I am here to offer my services if there is anything I can help with to make that a reality? Happy to submit what ever pull-request that I can, try to answer question that might exist and so on

Would that be something you'd be interested in?

**Comment by jeremymeng** [comment]
@JeremySkinner 
> Another question...does this need to be a separate project, or can we just pack the same binaries from the 4.6 PCL?

Yes, the same binaries should just work.  I created a new project simply because I don't want to change any of your existing code, while showing how to support netstandard1.3.

**Comment by JeremySkinner** [comment]
@jeremymeng thanks, that's what I thought but just wanted to clarify!

@thecodejunkie that would be very helpful and greatly appreciated if you have the time. The code changes needed for the core FV all to compile on coreclr are pretty much done, the main thing that will need doing is writing the MVC integration...but I'm guessing that's going to be out of your remit?

**Comment by thecodejunkie** [comment]
@JeremySkinner Sorry for the delayed reply

So I don't have any specifics on what and what I can't work on. It's more about trying to help get the artifacts out in sync with Microsofts stuff so that there is no, or as little as possible, delay on waiting for the Fluent Validation stuff to be available since that would block my client on completing their port to CoreCLR.

From a Nancy pov we really only reference the core package and create our own adapter between Nancys validation mechanism and Fluent Validation. So those are the primary bits of concern for us. That said, would you be willing to put out the core package before the MVC integration was done? I guess that is the big question here. I guess creating an epic issue, with a checklist of things that needs to be done for an rc2 release would be great and we can take it from there

Even if you were comfortable with putting out the code package first (and have the mvc integration follow after) then there are a couple of things you would need to do, because I think you'd also like to run your tests against core? Also, if I get it right, right now it's kind of "cheating" in the sense that it's taking the PCL build and sticking it into the correct folder in the nuget package? I.e it's not a proper convert using `project.json` and specifying the various TFMs that you support and so on?

I'll wait for your reply and take it from there.

Thanks

**Comment by JeremySkinner** [comment]
@thecodejunkie No objection to putting out a package with just the core FV dll....that was my original plan anyway, and then the MVC integration can follow afterwards.

It's not really "cheating" with the PCL build...the modern PCLs (4.6) use project.json too.

**Comment by thecodejunkie** [comment]
@JeremySkinner Cool. So I think we need to create a checklist issue with things that would be needed. Perhaps @jeremymeng also has some input here. For instance, running your tests against the CoreCLR (using the Dotnet Core requires some changes to your test projects, like supporting the `netcoreapp1.0` TFM for those and setting up the proper test hook so you can call `dotnet test`.

**Comment by thecodejunkie** [comment]
@JeremySkinner Oh just realized that my use of "we" might have been a bit misleading. Since you know best what the current state is (what works, what doesn't, what do you want to get working etc) for a release, I think it's good if you could draft the list and I can help enrich it

**Comment by JeremySkinner** [comment]
- Update core FluentValidation project to compile on coreclr (mostly done) 
- Add unit test project with linked test files to run tests against coreclr
- Add MVC6 integration project
- Make use of new (improved?) validation infrastructure in MVC6 to hook into the validation pipeline, rather than just porting the MVC5 integration.

**Comment by Kukkimonsuta** [comment]
I've got one more request for MVC6 integration and that would be to make it possible to use the new `IStringLocalizer` interface (or just the pattern to avoid direct dependency) for localization instead of static resources.

**Comment by jchannon** [comment]
Could @thecodejunkie help with setting up the unit test project?

**Comment by thecodejunkie** [comment]
@JeremySkinner @jeremymeng @jchannon with the announcement of the RC2 release-date, is it perhaps time for Meng to turn that into a proper PR so it can be reviewed and pulled asap so we can start looking into the other stuff? I can help with the test stuff, but there's no code to use right now :stuck_out_tongue:

**Comment by jeremymeng** [comment]
I submitted a PR to add .netstandard1.3 support by re-using the output from the PCL profile 259 project.
@thecodejunkie Since this is msbuild-based, and the story of referencing csproj from project.json isn't well supported.  You might need a wrap project for the assembly built from the PCL project.

**Comment by JeremySkinner** [comment]
Thanks...I'm going to hold off merging anything in until RC2 is actually out. It looks like there are still massive breaking changes happening, especially as David Fowler said this morning that project.json is goimg away.

**Comment by jeremymeng** [comment]
Sounds reasonable to wait for RC2 release. However, it is worth noting that my PR doesn't introduce any RC2 dependencies. It mostly just packs the same PCL build output for the `netstandard1.3` TFM.

**Comment by ddotlic** [comment]
> especially as David Fowler said this morning that project.json is going away.

I know this is somewhat unrelated to this particular discussion but... **WTH**? Where did you "hear" this? I knew there'll be breaking changes but not of _this_ magnitude.

**Comment by michaelvolz** [comment]
@ddotlic I found the tweet he mentioned: https://twitter.com/davidfowl/status/730219570783363073

**Comment by JeremySkinner** [comment]
It was announced last night in the latest Community Standup (video's online), and is on twitter too. I'm sure there will be some blog posts explaining the decision soon. 

The .NET core development process has been extremely disappointing....it's been a big FU to library developers since day 1, and I imagine that early adopters who are actually writing production apps will be extremely unhappy too.

**Comment by Kukkimonsuta** [comment]
According to the Comunity Standup `project.json` is NOT going away with RC2, most of the parts are already locked down for RC2 and it is expected to come out within next two weeks (probably sooner) and should be stable enough. The breaking change is that some tooling packages on the nightly nuget feeds (meaning this affects only people already working on RC2) get renamed with `bumped down` version (from `-rc2` to `preview1`) because they are splitting the release cycles of .NET Core / ASP.NET Core (which are in very mature state) and its tooling (which is very much new).

Changing `project.json` back to `.csproj` will happen around release of tooling `preview-2` which will happen around .NET Core / ASP.NET Core RTM release.

**Comment by jchannon** [comment]
If we could get @jeremymeng PR merged as it does not have Rc2 bits in it, and @thecodejunkie can start making a unit test PR for RC2 this won't hold us up when Rc2 ships

**Comment by JeremySkinner** [comment]
@jeremymeng thanks for clarifying...is there anything in the PR that you believe could potentially stop working between now and RC2? (Ie incompatibilities with prerelease version of nuget etc?)

**Comment by jeremymeng** [comment]
@JeremySkinner NuGet v3.4.3 in my PR is the latest release version.  No, I don't think anything would break in RC2.

**Comment by JeremySkinner** [comment]
Brilliant, thanks. I've merged in the PR.

**Comment by jchannon** [comment]
Awesome!

@JeremySkinner are you interested in getting a rc2 package out when its released if you have contributions or are you staying away from it until RTM and csproj stuff is sorted?

**Comment by ddotlic** [comment]
> are you staying away from it until RTM and csproj stuff is sorted?

From what I understood the `csproj` _stuff_ won't be sorted any time soon and certainly not at RTM (of the ASP.NET Core).

**Comment by JeremySkinner** [comment]
I'm in two minds at the moment. 

Part of me would like to get at least a prerelease package out with RC2 support (even if there's no MVC integration) as people have been asking for it for long enough, but on the other hand I have no faith that there won't be more significant breaking changes between RC2 and RTM and would rather wait until we're 100% sure what the final product is actually going to look like.

**Comment by ddotlic** [comment]
> there won't be more significant breaking changes between RC2 and RTM 

They repeatedly emphasize that there won't be any changes apart from bug fixes _for the library itself_.

The tooling will be in flux, but we don't generally care about that, no?

**Comment by jchannon** [comment]
Fingers crossed there wont be any API changes from RC2 to RTM, maybe @jeremymeng can provide any further info.

We as a company are wanting to get onto RC2 ASAP and have been funding Nancy to make that happen but we and Nancy are dependent on upstream packages.  We are ready to do a fork if necessary as we need to start testing our app but ideally we don't want to be in that position.  We would happily do any PRs if that helps persuade you?

**Comment by JeremySkinner** [comment]
@ddotlic, 

> They repeatedly emphasize that there won't be any changes apart from bug fixes for the library itself.
...and yet time and time again that's been proven not to be the case. 

> The tooling will be in flux, but we don't generally care about that, no?
I care, as I have to work with it when targetting core :)

**Comment by JeremySkinner** [comment]
@jchannon  Yeah, a fork is definitely far from idea....for now let's aim for having a prerelease package out pretty soon after RC2 is released, but it'll only be for the main assembly, with no MVC6 support (but you don't need that anyway). I'm hoping it'll just be a case of doing a rebuild and push to nuget.

**Comment by jchannon** [comment]
![](http://i.giphy.com/5n18DqTunxt4I.gif)

@JeremySkinner would you like help with the unit tests being compilable/runnable for coreclr before you release that rc2 package?

**Comment by JeremySkinner** [comment]
Yes, that would be great if possible, although I'm less bothered about that for a prerelease package. Definitely want it in time for RTM though!

**Comment by jchannon** [comment]
OK brill. Will see if I can persuade @thecodejunkie to look on Monday

Thanks again!

**Comment by jchannon** [comment]
or @jeremymeng if he's available 😉

**Comment by JeremySkinner** [comment]
What I've done so far with the tests:
- Upgraded all test projects to xunit 2.1 (Except the Portable40 build, which is still on 1.9.2)
- Introduced a test project for the Portable46 build (running on regular framework 4.6.1)
- Modified the build script to use xunit 2.1 

There seems to be an issue with xunit 2.1 where a test class implements IDisposable, Dispose isn't always called, meaning test cleanup isn't always being run. Not sure why this is happening as it worked fine in 1.9.2. I've tried to work around this, and all tests are now passing on my system, but please let me know if any of you run into any failing tests (especially related to localization). This only seems to apply to the msbuild-based runner- the resharper runner is fine.

**Comment by JeremySkinner** [comment]
**Notes for MVC 6 integration:**

- Validation should be handled via the IObjectModelValidator interface. This provides access to validate the entire model object *after* binding has occurred, which fits with FV's validation approach. This is much better than messing with ModelValidatorProviders and property-level validators that we had to do in MVC5 and earlier.
- We'll still run into the problem of the default binding infrastructure adding "Value is required" messages for non-nullable value types if no value is specified, although it should be possible to intercept this (not sure yet; will update when I know)
- Not looked at client side integration yet. Lowest priority.
- Tests for this should be full integration tests- fire up MVC hosted in-memory and excercise the full request pipeline to ensure tests are as robust as possible. 
- Use same tests scenarios as are in the current MVC integration

**Comment by jeremymeng** [comment]
@JeremySkinner Thanks for the xunit changes!  it makes adding .net core test easier.

> or @jeremymeng if he's available

I will be working on building the Core FV project against RC2, and adding a .NET core test project.

**Comment by jchannon** [comment]
@jeremymeng how we doing for a PR?

**Comment by JeremySkinner** [comment]
Aiming to merge it in later this week. Hopefully will have some time tomorrow.

**Comment by jchannon** [comment]
Awesome!!

**Comment by lucabriguglia** [comment]
Great!

**Comment by jchannon** [comment]
Will it do an auto deploy to a CI/alpha nuget feed or is that a manual step
a present?

**Comment by JeremySkinner** [comment]
I've merged it in, and I've deployed 6.3.3-alpha to nuget.

Please note I've only pushed up the main package for this release:
- There are no MVC5/WebApi packages
- There are no signed packages 

@jchannon There's no automated ci build/nuget deployment- I have a script for pushing up the nuget packages that I run locally when I want to do a release located in .build\publish-nuget-packages.ps1 

Please give this a try and report back on how it looks.

**Comment by jchannon** [comment]
Great work, pulled it and added it to Nancy.Validation.FluentValidation, one minor reflection change needed in Nancy.Validation.FluentValidation, ran a sample app and all worked perfectly!

Great job @JeremySkinner & @jeremymeng !

**Comment by lucabriguglia** [comment]
Thank you guys, this is the last missing peace to convert my application to .NET Core. I'll try it out later on today.

**Comment by waynebrantley** [comment]
@JeremySkinner Great news here!
(btw, maybe consider looking at appveyor for this project and let everything be automated.  It is free for open source.  Just a thought.)

**Comment by JeremySkinner** [comment]
I'm actually happy with the build process as it is at the moment- Thanks for the suggestion though!

**Comment by thecodejunkie** [comment]
@JeremySkinner it does not have to be an all-or-nothing CI setup.. we use AppVeyor to run our tests on core, Travis for Mono and TC for Windows 

![image](https://cloud.githubusercontent.com/assets/50543/15606817/d65cdafa-240e-11e6-94ae-b04f52b456db.png)

**Comment by lruckman** [comment]
Don't know if it's worthwhile but I just cobbled together a integration for MVC6. It's not feature complete but does seem to validate objects through the IObjectModelValidator. DefaultBinder messages of "the value is '' is invalid." still show however. If interested I could put it somewhere. Also includes registration extension hanging off of IMvcBuilder so it's easy to add in Start.cs.

**Comment by JeremySkinner** [comment]
Sounds good...if you want to throw up some the code somewhere then I'll
take a look at it. Thanks!

**Comment by lruckman** [comment]
@JeremySkinner The three required files are [here](https://github.com/lruckman/DRS/tree/master/src/Web/Engine/Validation).

Adding it to MVC takes place [here](https://github.com/lruckman/DRS/blob/master/src/Web/Startup.cs) on line 67

The code when adding FluentValidation clears the other ModelValidationProviders.

**Comment by lruckman** [comment]
Looks like I have some issue with it, mainly everything's invalid even with no errors. Need to look into getting the IsValid set to true when the validation is successful. I'll look into it further.

**Comment by buvinghausen** [comment]
@lruckman Larry thank you for sharing that.  There were two issues that kept it from working for me but I have gotten them fixed and everything seems to be playing nice.

1) The Assembly property is no longer on the Type class in .NET Core you can however call to .GetTypeInfo() from the Type which then does have the Assembly property

2) I needed the matching interface for IValidator<t> in order to pass both parameters into mvcBuilder.Services.AddTransient so I basically created a hybrid of what you were doing with the AssemblyScanner from the main branch and came up with this

`typeof(FluentValidationObjectModelValidator)
	.GetTypeInfo().Assembly.GetTypes()
	.Select(t => new
	{
		ValidatorType = t,
		InterfaceType = t.GetInterfaces().FirstOrDefault(i => i.GetTypeInfo().IsGenericType && i.GetGenericTypeDefinition() == openGenericType)
	})
	.Where(t => t.InterfaceType != null)
	.ToList().ForEach(f =>
	{
		mvcBuilder.Services.AddTransient(f.InterfaceType, f.ValidatorType);
	});`

**Comment by JeremySkinner** [comment]
Question for @jeremymeng, is there any reason we need to keep the Net46 PCL now we have the NetStandard version?

**Comment by JeremySkinner** [comment]
@lruckman Thanks again for the prototype- I'm going to try and find some time next week to start fleshing it out.

**Comment by jeremymeng** [comment]
> is there any reason we need to keep the Net46 PCL now we have the NetStandard version?

@JeremySkinner Good question. I think it's a matter of whether pre-release dependencies should be used.  At the moment Net46 PCL dependends on released (older) version of packages, and its output is used by the uap10.0 TFM. NetStandard version still depends on RC2 bits.  With the incoming RTM, and newer nuget client that supports VS 2012/2013, I believe we can just have netstandard project after upgrade it to use RTM bits.

**Comment by JeremySkinner** [comment]
@jeremymeng Thanks, that's what I was thinking too. I'll keep them both around until RTM. Are you also also expecting the need to wrap the Netstandard PCL to disappear?

**Comment by jchannon** [comment]
RTM is 2 days away don't forget :)

**Comment by JeremySkinner** [comment]
So I've done some more work on this today:

- Initial MVC6 support based on @lruckman's work.
- Ported some of the MVC5 tests over to MVC6 (most currently failing) 

What's left to do:
- Fix the MVC6 tests. 
- Add support for CustomizeValidator and IValidationInterceptor
- Clientside support at some point in the future 
- Build script modifications to create FluentValidation.Mvc6 nuget package (someone else already has this registered on nuget, so am waiting to get ownership transferred over)

**Comment by lruckman** [comment]
So I just recently found that with the example I provided if a action recieves a parameter like a string and not and object that we are validating through fluent validation it will leave the modelstate invalid because that argument doesn't ever get validated since I stripped all other validations from the pipeline in the registration helper. My workaround was simple, move the parameter into a validated object but it's not ideal for others.

**Comment by JeremySkinner** [comment]
It should be possible to fall back to the default implementation. I'll see if I can get that working.

**Comment by JeremySkinner** [comment]
@lruckman @buvinghausen just a note registering validators, my plan is to add a configuration expression to the AddFluentValidationCall:

```
builder.AddFluentValidation(fv => {
  cfg.RegisterValidatorsFromAssemblyContaining<MyValidator>();
});
```

...this will register these validators with the container. Alternatively, you can specify a ValidatorFactory if you want to use a custom factory. 

```
builder.AddFluentValidation(fv => {
  cfg.ValidatorFactory = new AttributedValidatorFactory();
});
```
Note that validators should be registered with AddSingleton and not AddTransient, and this is what RegisterValidatorsFromAssemblyContaining will do internally.

**Comment by JeremySkinner** [comment]
Good morning,

I've just pushed out FV 6.4 Beta 1 with support for .NET core 1.0 and ASP.NET Core. A few notes:

- All dependencies should now be up to date with net core 1.0. Let me know if you find any I've missed.
- The MVC integration is in the package "FluentValidation.AspNetCore" for consistency with Microsoft's naming convention
- The FluentValidation.MVC6 package is just an alias for FluentValidation.AspNetCore. 
- I've removed the wrapper project for FV.NetStandard.

Please give these packages a try and if you find any problems please open a new issue rather than commenting on this one. 

A massive thankyou to everyone who contributed time to making this work. 

Ps, If anyone wants to help with the documentation, please let me know ;)

**Comment by jeremymeng** [comment]
> Are you also also expecting the need to wrap the Netstandard PCL to disappear?

At the moment not yet as far as I know.

**Comment by lruckman** [comment]
The FluentValidation.AspNetCore package doesn't seem to like .NET Core net461 projects.

>Package FluentValidation.AspNetCore 6.4.0-beta2 is not compatible with net461 (.NETFramework,Version=v4.6.1). Package FluentValidation.AspNetCore 6.4.0-beta2 supports: netstandard1.6 (.NETStandard,Version=v1.6)
    One or more packages are incompatible with .NETFramework,Version=v4.6.1.

**Comment by JeremySkinner** [comment]
I'll get that fixed

**Comment by JeremySkinner** [comment]
I've pushed out a beta3 package- please give it another try.

**Comment by JeremySkinner** [comment]
@jeremymeng I'm seeing an issue with universal windows apps having versioning problems. Could this be related to the specific dependencies we added into the FluentValidation.nuspec? These are the kind of errors I'm seeing on an install-package fluentvalidation:

`Detected package downgrade: System.Runtime.WindowsRuntime from 4.0.11 to 4.0.10 
 App2 (>= 1.0.0) -> Microsoft.NETCore.UniversalWindowsPlatform (>= 5.1.0) -> Microsoft.NETCore (>= 5.0.0) -> System.IO.FileSystem (>= 4.0.0) -> runtime.win.System.IO.FileSystem (>= 4.0.1) -> System.Runtime.WindowsRuntime (>= 4.0.11) 
 App2 (>= 1.0.0) -> Microsoft.NETCore.UniversalWindowsPlatform (>= 5.1.0) -> System.Runtime.WindowsRuntime (>= 4.0.10)`
...etc

`System.Reflection.Emit.ILGeneration 4.0.1 provides a compile-time reference assembly for System.Reflection.Emit.ILGeneration on UAP,Version=v10.0, but there is no run-time assembly compatible with win10-arm-aot.`

Is that something you've seen before?

**Comment by lruckman** [comment]
@JeremySkinner that works, thanks a bunch!

**Comment by jeremymeng** [comment]
> I'm seeing an issue with universal windows apps having versioning problems. Could this be related to the specific dependencies we added into the FluentValidation.nuspec?

I think for UAP, NET Core platform 5.2.0 is needed to support new .NET Core bits.  I used NuSpec.ReferenceGenerator to update the dependencies in the nuspec file, which automatically uses the compatible (same major and minor, higher revision) versions. I tested and verify that beta3 package works with `"Microsoft.NETCore.UniversalWindowsPlatform": "5.2.0"`.

Since we are using the portable binaries for profile259 and uap10.0, it might be better to use the original dependency versions.  I submitted https://github.com/JeremySkinner/FluentValidation/pull/294. It worked with `"Microsoft.NETCore.UniversalWindowsPlatform": "5.1.0"` in my local testing.

**Comment by JeremySkinner** [comment]
Thanks, I'll try and get this merged in this week.

**Comment by JeremySkinner** [comment]
That's done- thanks.

**Comment by arielcloud** [comment]
Where can I found any example or docs about injecting validators in asp.net built-in DI?
I saw [this](https://github.com/JeremySkinner/FluentValidation/wiki/i.-IoC), but is for structureMap. thanks!

**Comment by JeremySkinner** [comment]
@arielcloud https://docs.asp.net/en/latest/fundamentals/dependency-injection.html

**Comment by arielcloud** [comment]
I not see any "TryResolve" or "TryGetInstance" in https://docs.asp.net/en/latest/fundamentals/dependency-injection.html 
so, how can I create the "ValidatorFactory". if you can give more details - it will be great.

for my services I just create an interface, and inject it to the DI by for example: `services.AddTransient<IModelValidation, ModelValidation>();`

but with fluent validation I don't understand how to use it.

**Comment by JeremySkinner** [comment]
I'm not sure I understand what you're asking. 

You register your validators with the container in the same way as you described:

`services.AddSingleton<IValidator<Person>, PersonValidator>();`

You don't need to write a validator factory. The default one will handle resolution from the container.

**Comment by hmvs** [comment]
Would be nice to have extension for IMvcCoreBuilder
to have ability to call services.AddMvcCore().AddFluentValidation()

**Comment by JeremySkinner** [comment]
I'll see about getting that added....I didn't even know there was an AddMvcCore! Seems quite confusing to have 2 entry points for configuration

**Comment by arielcloud** [comment]
I got 

> 'IMvcBuilder' does not contain a definition for 'AddFluentValidation' and no extension method 'AddFluentValidation' accepting a first argument of type 'IMvcBuilder' could be found (are you missing a using directive or an assembly reference?)	

I dont missed `using FluentValidation;` in my startup.cs file, and I have `"FluentValidation": "6.4.0-beta3",`

**Comment by JeremySkinner** [comment]
You need to have installed the FluentValidation.AspNetCore package and have a `using FluentValidation.AspNetCore`.

**Comment by gdoron** [comment]
Regarding `AddMvcCore`:

https://github.com/aspnet/Mvc/issues/2872

**Comment by hmvs** [comment]
@JeremySkinner saw that you committed  `AddFluentValidation` for the `IMvcCoreBuilder`. Thanks for that :) Would be nice to have it in nuget.

**Comment by rono1** [comment]
I noticed this is still listed as prerelease in nuget.  Is this basically stable enough to use in production, or are there still pending issues?

**Comment by CumpsD** [comment]
Just trying to summarize here, to see what should be working at this stage, please correct me where I'm missing something:

 * Latest version: `"FluentValidation.AspNetCore": "6.4.0-beta3"`

In `Startup` the only change needed is:

```
services
    .AddMvc()
    .AddFluentValidation(fv => fv.RegisterValidatorsFromAssemblyContaining<Startup>());
```

Types are decorated with `Validator`

```
[Validator(typeof(CreateLocationRequestValidator))]
public class CreateLocationRequest
{
    public int? ExternalId { get; set; }

    public Address Address { get; set; }
}
```

Validators are made with `AbstractValidator<T>`
```
public class CreateLocationRequestValidator : AbstractValidator<CreateLocationRequest>
{
    public CreateLocationRequestValidator()
    {
        RuleFor(x => x.ExternalId)
            .NotEmpty()
            .When(IsAddressEmpty)
            .WithMessage("ExternalId or Address is required.");

        RuleFor(x => x.Address)
            .NotEmpty()
            .When(x => !x.ExternalId.HasValue)
            .WithMessage("ExternalId or Address is required.");
    }
```

After doing all of the above, this should work?
```
        [HttpPost]
        public ActionResult Post([FromBody] CreateLocationRequest message)
        {
            if (!ModelState.IsValid)
                return BadRequest();

            return Ok();
        }
```

I think I'm missing something, since my validator isn't being hit in debug ;) But what am I missing exactly?

**Comment by JeremySkinner** [comment]
@CumpsD 

Calling RegisterValidatorsFromAssemblyContaining... registers the validators with MVC's built-in IoC container, meaning you should remove the ValidatorAttribute from your validators. The ValidatorAttribute is for use with the AttributedValidatorFactory, which was the default way of instantating validators in older versions of MVC. You can still use this if you want, but you'd need to set this up in your AddFluentValidation call. 

If you run into further problems, please open a separate issue. Thanks!

**Comment by CumpsD** [comment]
Thanks for the clarification 👍

**Comment by JeremySkinner** [comment]
For those still following the progress on this, I'm hoping to have support for clientside validation on asp.net core done by next week, with an updated beta package shortly after.

**Comment by JeremySkinner** [comment]
6.4 beta10 is out with clientside support. Please give it a try. If this all looks good then this is the build that'll go out as "stable".

**Comment by Usman235** [comment]
@CumpsD i did the same thing which you mention above (Code) its working fine . let me explain my project structure first .
![Project Structure](https://cloud.githubusercontent.com/assets/14319269/22588310/0b3368b4-ea27-11e6-9f96-d081dc2c41dc.PNG)

Above you can see the project structure of mine . if i put my following code in figure section-1(FH.Common) it not validating the model . but if i put i asp.net core web api (FH.Webapi) solution its working fine .

```
 [Validator(typeof(CurrencyDTOValidator))]
    public class CurrencyDTO
    {
        public int Id { get { return CurrencyId; } }
        public int CurrencyId { get; set; }
        public string Name { get; set; }
        public string Symbol { get; set; }

    }*/

    public class CurrencyDTOValidator : AbstractValidator<CurrencyDTO>
    {
        public CurrencyDTOValidator()
        {
            RuleFor(x => x.Name).NotEmpty().NotNull().WithMessage("The currency 'Name' is required.")
                                      .Length(0, 250).WithMessage("The currency 'Name' cannot be more than 250 characters.");

            RuleFor(x => x.Symbol).NotEmpty().WithMessage("The currency Symbol is required.");

            //RuleFor(x => x.BirthDate).LessThan(DateTime.Today).WithMessage("You cannot enter a birth date in the future.");
        
            //RuleFor(x => x.Username).Length(8, 999).WithMessage("The user name must be at least 8 characters long.");
        }
    }
```


Now may question is that is there any other thing i have to do to run it through classlibrary

**Comment by JeremySkinner** [comment]
FV 6.4 is out with the final version of the aspnetcore integration. Thanks everyone for the help and feedback. If you run into any problems, please let me know on separate issues.
